sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DistributionAndOrderingUtils.scala

```diff
@@ -36,16 +36,27 @@ object DistributionAndOrderingUtils {
         case _: UnspecifiedDistribution => Array.empty[Expression]
       }
 
+      val (sortOrder, _) = distribution.partition(_.isInstanceOf[SortOrder])
+
       val queryWithDistribution = if (distribution.nonEmpty) {
-        val finalNumPartitions = if (numPartitions > 0) {
-          numPartitions
+        // Spark can optimize the partition when
+        // 1. numPartitions is not specified by the data source, and
+        // 2. sortOrder is specified. This is because the requested distribution needs to be
+        // guaranteed, which can only be achieved by using RangePartitioning, not HashPartitioning.
```

Comment:
This looks too strict. Do we have a reason to must keep the hash partitioning? For file sources, we want to shuffle by partition columns to reduce the number of files, but it's just a best effort as a rough hash partitioning also works.

Do we have other special requirements for DS v2?

Reply:
Anyway, we need a DS v2 API to let the sink require "rough" hash partitioning. If we don't want to change the existing API, we need to add a new one for it.